mlir/lib/Catalyst/Transforms/QnodeToAsyncPatterns.cpp

```diff
@@ -0,0 +1,101 @@
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+namespace catalyst {
+
+struct CallOpToAsyncOPRewritePattern : public mlir::OpRewritePattern<func::CallOp> {
+    using mlir::OpRewritePattern<func::CallOp>::OpRewritePattern;
+
+    mlir::LogicalResult matchAndRewrite(func::CallOp op,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        SymbolRefAttr symbol = dyn_cast_if_present<SymbolRefAttr>(op.getCallableForCallee());
+        func::FuncOp func = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, symbol);
+        if (!func) {
+            // Nothing to change.
+            return failure();
+        }
+
+        // Check for Call ops that have QNode func ops
+        if (!func->hasAttrOfType<UnitAttr>("qnode")) {
+            // Nothing to change. (For functions which are not qnodes).
+            return failure();
+        }
+
+        if (op->hasAttrOfType<UnitAttr>("transformed")) {
+            // Nothing to change. (For functions which are not qnodes).
+            return failure();
+        }
+
+        TypeRange retTy = op.getResultTypes();
+        SmallVector<Value> dependencies; /* = empty */
+        SmallVector<Value> operands;     /* = empty */
+        auto noopExec = [&](OpBuilder &executeBuilder, Location executeLoc,
+                            ValueRange executeArgs) {};
+
+        rewriter.updateRootInPlace(op, [&] { op->setAttr("transformed", rewriter.getUnitAttr()); });
+        IRMapping map;
+        auto executeOp =
+            rewriter.create<async::ExecuteOp>(op.getLoc(), retTy, dependencies, operands, noopExec);
+        {
+            PatternRewriter::InsertionGuard insertGuard(rewriter);
+            rewriter.setInsertionPoint(executeOp.getBody(), executeOp.getBody()->end());
+            Operation *cloneOp = op->clone(map);
+            rewriter.insert(cloneOp);
+            rewriter.create<async::YieldOp>(op.getLoc(), cloneOp->getResults());
+        }
+
+        auto asyncValues = executeOp.getResults();
+
+        // We should move this as much as we can...
+        rewriter.create<async::AwaitOp>(op.getLoc(), asyncValues.front());
+        rewriter.create<async::RuntimeDropRefOp>(op.getLoc(), asyncValues.front(),
+                                                 rewriter.getI64IntegerAttr(1));
+
+        std::vector<Value> bodyReturns(asyncValues.begin() + 1, asyncValues.end());
+        if (bodyReturns.size() > 0) {
+            for (auto [oldVal, newVal] : llvm::zip(op.getResults(), bodyReturns)) {
+                auto awaitOp = rewriter.create<async::AwaitOp>(op.getLoc(), newVal);
+                rewriter.replaceAllUsesWith(oldVal, awaitOp.getResults());
+                rewriter.create<async::RuntimeDropRefOp>(op.getLoc(), newVal,
```

Comment:
What does the runtime drop ref does exactly?

Reply:
[From MLIR's documentation](https://mlir.llvm.org/docs/Dialects/AsyncDialect/#asyncruntimedrop_ref-asyncruntimedroprefop)

> Drops a reference to async value

Looking at the implementation of AsyncExecutionEngine we see the following comments:

```
  // AsyncToken created with a reference count of 2 because it will be returned
  // to the `async.execute` caller and also will be later on emplaced by the
  // asynchronously executed task. If the caller immediately will drop its
  // reference we must ensure that the token will be alive until the
  // asynchronous operation is completed.
  AsyncToken(AsyncRuntime *runtime)
      : RefCounted(runtime, /*refCount=*/2), state(State::kUnavailable) {}

  // AsyncValue similar to an AsyncToken created with a reference count of 2.
  AsyncValue(AsyncRuntime *runtime, int64_t size)
```

The value being dropped appears to be just a pointer to the memref struct.